KAFKA-3112: Change verification of bootstrap servers

bondj opened the pull request:
so that unresolvable DNS names are ignored and only throw an error if no other bootstrap servers are resolvable.

JonathanBond committed:
Change verification of DNS names to only throw an error if no bootstrap
servers are addressable.

bondj commented:
The contribution is my original work and that I license the work to the project under the project's open source license.

granthenke commented:
Thanks for the patch. Could you change the title of the PR to lead with "KAFKA-3112: "? A unit test would be great too.

JonathanBond committed:
Add unit tests

bondj commented:
Done & done.  There's also a test in KafkaProducerTest which goes through this code.

granthenke commented:
@bondj Thanks for the update. 

One thing to note is that any invalid url will no longer be a part of the bootstrap, even if it is resolvable later because most usage is something like this:

``` scala
List<InetSocketAddress> addresses = ClientUtils.parseAndValidateAddresses(config.getList(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG));
this.metadata.update(Cluster.bootstrap(addresses), 0);
```

I would like some others to confirm thats acceptable. @ewencp @ijuma ?

The name of the method `parseAndValidateAddresses` doesn't make it totally clear that it could filter out and return a subset of addresses, though its purpose is clearly for bootstrap use given the logging inside the method and its current usage in the project. I think some simple javadoc on the behavior of the method would be useful. 

Otherwise LGTM.

granthenke commented:
Also noting this change looks to be very similar/duplicate to [KAFKA-2657](https://issues.apache.org/jira/browse/KAFKA-2657) with #336. Closing one should resolve the other. I have linked them in Jira.

ijuma commented:
Good point with regards to #336 @granthenke. It would be good to understand why unresolvable DNS names are present and whether this is a temporary failure or a permanent one. It does seem a bit odd to remove it from the list if it's just a temporary failure.

ewencp commented on clients/src/main/java/org/apache/kafka/clients/ClientUtils.java:
```diff
@@ -44,16 +44,18 @@
-                        throw new ConfigException("DNS resolution failed for url in " + CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG + ": " + url);
-                    addresses.add(address);
+                    if (address.isUnresolved()) {
+                        log.warn("DNS resolution failed for url in " + CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG + ": " + url);
```
Maybe it would help clarify the behavior if this warning also indicated it is being removed from the bootstrap server list?

ewencp commented on clients/src/test/java/org/apache/kafka/clients/ClientUtilsTest.java:
```diff
@@ -29,12 +29,18 @@ public void testParseAndValidateAddresses() {
         check("mydomain.com:8080");
         check("[::1]:8000");
         check("[2001:db8:85a3:8d3:1319:8a2e:370:7348]:1234", "mydomain.com:10000");
+        check("some.invalid.hostname.foo.bar:9999", "mydomain.com:10000");
```
Given that this is now supposed to return a specific subset, you might want to update `check()` to return the list so the tests can validate the output.

bondj commented:
@ijuma My understanding is the modified bootstrap servers list is only used during the initialisation and constructor of the client.  Once used to bootstrap, the node information is all replaced with the current metadata, and then refreshed periodically. 

The benefit of the fix is that it allows the client to start if any one of the bootstrap servers are resolvable. We might have a connectivity problem to a single node when we actually start to send messages, but this can be fixed through the usual refresh cycle of the metadata as and when the system as a whole becomes aware of the issue. 

It reduces the impact of a single bad DNS record from preventing the client to start, to bad connectivity to a single broker, until the DNS problem is resolved - or the metadata is updated to include a new broker with a better DNS entry.

Preventing the startup of the client because of one bad dns entry when there are others which are usable seems overly restrictive.  If there are no servers resolvable the client should be prevented from starting and throwing the relevant exception.

ijuma commented:
@bondj There is a discussion about using the bootstrap servers after initialisation under some failure scenarios in KAFKA-3068.

ewencp commented:
@ijuma I think this approach is a reasonable compromise given that we'd want to exclude the bad hostnames from the Cluster. Maybe the right solution is to do this (which has no negative affect re: KAFKA-3068 given current code doesn't actually replace with the bootstrap servers, just keeps track of seen nodes). Then KAFKA-3068 can address whether the full original list is re-loaded or we use the filtered version based on this code when it switches to reverting to the original bootstrap servers list.

ijuma commented:
@ewencp fine by me.

granthenke commented:
@ewencp Should this go in now then, with a follow up in KAFKA-3068?
Note: It should also close KAFKA-2657.

bondj commented:
@granthenke @ewencp @ijuma will this be included in to the next release of kafka?

ijuma commented:
@bondj, the gist of the change seems fine. Can you please address @ewencp's two comments? I'll merge it to trunk once you do.

JonathanBond committed:
Incorporate review comments

Change message to highlight server removed from bootstrap servers
Verify only good server left in validated list of servers

JonathanBond committed:
Fixing spacing on unit test

bondj commented:
@ijuma done!

ijuma commented:
LGTM, merged to trunk and 0.10.0 branches.
